File: loader/src/main/java/de/timolia/lactea/loader/inject/InjectedInstance.java
package de.timolia.lactea.loader.inject;

import com.google.inject.Injector;
import java.util.Objects;
import java.util.function.Function;

/**
 * @author David (_Esel)
 */
public interface InjectedInstance<T> {
    static <T> InjectedInstance<T> ofInstance(T instance) {
        return injector -> instance;
    }

    static <T> InjectedInstance<T> ofClass(Class<? extends T> clazz) {
        Objects.requireNonNull(clazz, "class");
        return injector -> injector.getInstance(clazz);
    }

    static <T, F> InjectedInstance<T> ofFactory(Class<? extends F> factoryClass,
                                                Function<F, T> createFunction) {
        Objects.requireNonNull(factoryClass, "factoryClass");
        Objects.requireNonNull(createFunction, "createFunction");
        return injector -> createFunction.apply(injector.getInstance(factoryClass));
    }

    T getInstance(Injector injector);
}

File: module-api/src/main/kotlin/de/timolia/lactea/inject/InjectedInstance.kt
package de.timolia.lactea.inject

import com.google.inject.Injector

fun interface InjectedInstance<T> {
    fun getInstance(injector: Injector): T

    companion object {
        fun <T> ofInstance(instance: T): InjectedInstance<T> {
            return InjectedInstance { instance }
        }

        fun <T> ofClass(clazz: Class<out T>): InjectedInstance<T> {
            return InjectedInstance { injector -> injector.getInstance(clazz) }
        }

        fun <T, F> ofFactory(
            factoryClass: Class<out F>,
            createFunction: (F) -> T
        ): InjectedInstance<T> {
            return InjectedInstance { injector -> createFunction(injector.getInstance(factoryClass)) }
        }
    }
}